python_modules/dagster/dagster/_core/definitions/graph_definition.py

```diff
@@ -859,42 +856,45 @@ def _validate_in_mappings(
+                    f'"{maps_to.node_name}.{maps_to.input_name}" index {maps_to.fan_in_index} in '
                     f"the MultiDependencyDefinition is not a MappedInputPlaceholder"
                 )
-            mapping_keys.add(f"{maps_to.solid_name}.{maps_to.input_name}.{maps_to.fan_in_index}")
+            mapping_keys.add(f"{maps_to.node_name}.{maps_to.input_name}.{maps_to.fan_in_index}")
         else:
             if dependency_structure.has_deps(solid_input_handle):
                 raise DagsterInvalidDefinitionError(
-                    "In {class_name} '{name}' input mapping target "
-                    '"{mapping.maps_to.solid_name}.{mapping.maps_to.input_name}" '
-                    "is already satisfied by output".format(
-                        name=name, mapping=mapping, class_name=class_name
-                    )
+                    f"In {class_name} '{name}' input mapping target "
+                    f'"{mapping.maps_to.node_name}.{mapping.maps_to.input_name}" '
+                    "is already satisfied by output"
                 )
 
-            mapping_keys.add(
-                "{mapping.maps_to.solid_name}.{mapping.maps_to.input_name}".format(mapping=mapping)
-            )
+            mapping_keys.add(f"{mapping.maps_to.node_name}.{mapping.maps_to.input_name}")
 
     for input_handle in dependency_structure.input_handles():
         if dependency_structure.has_fan_in_deps(input_handle):
             for idx, dep in enumerate(dependency_structure.get_fan_in_deps(input_handle)):
                 if dep is MappedInputPlaceholder:
-                    mapping_str = (
-                        "{input_handle.solid_name}.{input_handle.input_name}.{idx}".format(
-                            input_handle=input_handle, idx=idx
-                        )
-                    )
+                    mapping_str = f"{input_handle.node_name}.{input_handle.input_name}.{idx}"
                     if mapping_str not in mapping_keys:
                         raise DagsterInvalidDefinitionError(
-                            "Unsatisfied MappedInputPlaceholder at index {idx} in "
-                            "MultiDependencyDefinition for '{input_handle.solid_name}.{input_handle.input_name}'".format(
-                                input_handle=input_handle, idx=idx
-                            )
+                            f"Unsatisfied MappedInputPlaceholder at index {idx} in "
+                            f"MultiDependencyDefinition for '{input_handle.node_name}.{input_handle.input_name}'"
                         )
 
-    return input_mappings, list(input_def_dict.values())
+    # if the dagster type on a graph input is None and all its target inputs have the
+    # same dagster type, then use that dagster type for the graph input
+    for graph_input_name, graph_input_def in input_defs_by_name.items():
+        if graph_input_def.dagster_type.kind == DagsterTypeKind.ANY:
```

Comment:
we have a `_type_not_set` bool on `InputDefinition` you can use to check for unset type. The case of concern would be an explicit `Any`

Reply:
I actually tried that first, but it didn't work for some reason. I think it's not making it through due to some sequence of `combine_with_inferred` calls or something, and I was hesitant to interrupt that delicate ecosystem.

My thinking was basically that it would be fine to override explicit Any. If I understand correctly, the only consequence ares that:
- If all the inner types are Nothing, then you don't need an upstream output, which I think should be true independent of the outer type.
- If all the inner types are the same Dagster type, then you can use the type loader on that type, which I think should be true independent of the outer type.

Anything I'm missing here?

Reply:
I cant think of anything, just update the comment `if the dagster type on a graph input is None` to match then.